libraries/bsg_manycore.cpp

```diff
@@ -1362,38 +1363,62 @@ static int hb_mc_manycore_recv_read_rsp(hb_mc_manycore_t *mc,
                 return err;
         }
 
-        /* check that the npa matches? */
-        
-        int shift = CHAR_BIT * (hb_mc_npa_get_epa(npa) & 0x3);
         /* read data from packet */
-        switch (sz) {
-        case 4:
-                *(uint32_t*)vp = hb_mc_response_packet_get_data(&rsp.response);
-                break;
-        case 2:
-                *(uint16_t*)vp = (hb_mc_response_packet_get_data(&rsp.response) >> shift) & 0xFFFF;
-                break;
-        case 1:
-                *(uint8_t*)vp  = (hb_mc_response_packet_get_data(&rsp.response) >> shift) & 0xFF;
-                break;
-        default:
-                return HB_MC_INVALID;
-        }
+        *vp = hb_mc_response_packet_get_data(&rsp.response);
+
         if (id != nullptr)
                 *id = hb_mc_response_packet_get_load_id(&rsp.response);
 
         return HB_MC_SUCCESS;
 }
+
+template<typename UINT>
+static UINT hb_mc_manycore_mask_load_data(const hb_mc_npa_t *npa, uint32_t load_data)
+{
+        int shift = CHAR_BIT * (hb_mc_npa_get_epa(npa) & 0x3);
+        uint32_t result;
+
+        /* make sure this template is being used only as intended */
+        static_assert(std::is_unsigned<UINT>::value,
+                      "hb_mc_manycore_mask_load_data: UINT must be uint8_t, uint16_t, or uint32_t");
+
+        static_assert(std::is_integral<UINT>::value,
+                      "hb_mc_manycore_mask_load_data: UINT must be uint8_t, uint16_t, or uint32_t");
+
+        static_assert(sizeof(UINT) == 1 || sizeof(UINT) == 2 || sizeof(UINT) == 4,
+                      "hb_mc_manycore_mask_load_data: UINT must be uint8_t, uint16_t, or uint32_t");
+
+        if (sizeof(UINT) == 4) {
+                result = load_data;
+        } else if (sizeof(UINT) == 2) {
+                result = (load_data >> shift) & 0xFFFF;
+        } else if (sizeof(UINT) == 1) {
+                result = (load_data >> shift) & 0xFF;
+        }
```

Comment:
if you are loading half, and the byte offset is not half aligned, it should ignore the LSB to make it half-aligned?
Otherwise, for load-half across the word boundary, you need to do loads to support

Reply:
Currently the comments that document the interface are pretty explicit that loads need to be aligned to their size. The buffered read function only supports reading word aligned/sized buffers and it explicitly checks those conditions and returns an error if they are not met.